File: NoteBase/Core/NoteId.cs
using System;
using System.IO;
using System.Text.RegularExpressions;

namespace NoteBase.Core
{
    /// <summary>
    /// ノート ID 生成・slug 化のユーティリティ。
    /// 形式: yyyyMMdd-HHmmss-&lt;slug&gt; （衝突時は末尾にミリ秒や連番を付与）
    /// </summary>
    public static class NoteId
    {
        private static readonly Regex AsciiSlugChars =
            new Regex(@"[^a-z0-9]+", RegexOptions.Compiled);

        public static string Generate(string title, DateTime now)
        {
            var stamp = now.ToString("yyyyMMdd-HHmmss");
            var slug = Slugify(title);
            return stamp + "-" + slug;
        }

        public static string Slugify(string title)
        {
            if (string.IsNullOrEmpty(title)) return "untitled";
            var lower = title.ToLowerInvariant();
            var slug = AsciiSlugChars.Replace(lower, "-").Trim('-');
            if (string.IsNullOrEmpty(slug)) return "untitled";
            if (slug.Length > 60) slug = slug.Substring(0, 60).TrimEnd('-');
            if (string.IsNullOrEmpty(slug)) return "untitled";
            return slug;
        }

        /// <summary>
        /// parentDir 以下に同名フォルダが存在する場合、衝突回避した ID を返す。
        /// </summary>
        public static string EnsureUnique(string id, string parentDir, DateTime now)
        {
            if (!Directory.Exists(Path.Combine(parentDir, id))) return id;

            var ms = now.Millisecond.ToString("D3");
            var candidate = id + "-" + ms;
            if (!Directory.Exists(Path.Combine(parentDir, candidate))) return candidate;

            for (int i = 2; i < 1000; i++)
            {
                candidate = id + "-" + ms + "-" + i;
                if (!Directory.Exists(Path.Combine(parentDir, candidate)))
                    return candidate;
            }
            throw new InvalidOperationException("Failed to generate unique note id: " + id);
        }

        /// <summary>
        /// id の slug 部分を新しい slug に差し替える。
        /// 例: 20260428-090000-untitled → 20260428-090000-new-slug
        /// </summary>
        public static string ReplaceSlug(string id, string newTitle)
        {
            var parts = id.Split('-');
            if (parts.Length < 3) return id;
            return parts[0] + "-" + parts[1] + "-" + Slugify(newTitle);
        }
    }
}

File: notebase/core/note_id.py
from __future__ import annotations

import re
from datetime import datetime
from pathlib import Path

_ASCII_SLUG_RE = re.compile(r"[^a-z0-9]+")


def generate(title: str, now: datetime) -> str:
    """yyyyMMdd-HHmmss-<slug> 形式のノート ID を生成する。"""
    stamp = now.strftime("%Y%m%d-%H%M%S")
    return f"{stamp}-{slugify(title)}"


def slugify(title: str | None) -> str:
    """タイトルを slug に変換する。

    規則:
    - 小文字化
    - ASCII 英数字以外の連続を '-' 1 文字に置換
    - 両端の '-' をトリム
    - 空文字なら "untitled"
    - 60 文字超なら 60 で切り詰め、末尾の '-' を再トリム
    """
    if not title:
        return "untitled"
    lower = title.lower()
    slug = _ASCII_SLUG_RE.sub("-", lower).strip("-")
    if not slug:
        return "untitled"
    if len(slug) > 60:
        slug = slug[:60].rstrip("-")
    if not slug:
        return "untitled"
    return slug


def ensure_unique(note_id: str, parent_dir: str | Path, now: datetime) -> str:
    """parent_dir 配下に同名フォルダがある場合、衝突回避した ID を返す。

    既存と被らない形になるまで以下の順で試す:
      1. 元の note_id
      2. note_id-<ms> (ミリ秒 3 桁)
      3. note_id-<ms>-2, note_id-<ms>-3, ... (最大 999)
    """
    parent = Path(parent_dir)
    if not (parent / note_id).exists():
        return note_id

    ms = f"{now.microsecond // 1000:03d}"
    candidate = f"{note_id}-{ms}"
    if not (parent / candidate).exists():
        return candidate

    for i in range(2, 1000):
        candidate = f"{note_id}-{ms}-{i}"
        if not (parent / candidate).exists():
            return candidate
    raise RuntimeError(f"Failed to generate unique note id: {note_id}")


def replace_slug(note_id: str, new_title: str | None) -> str:
    """ID 末尾の slug を new_title から再計算した値に差し替える。

    parts が 3 未満の場合は変更しない (壊れた ID 保護)。
    """
    parts = note_id.split("-")
    if len(parts) < 3:
        return note_id
    return f"{parts[0]}-{parts[1]}-{slugify(new_title)}"
